### ISSUE TITLE
Static Analysis for Identifying Data Parallelism Using LLVM/Clang

### ISSUE THREAD

**Issue by HITESH-S-P** [llvm/llvm-project#190508]
Develop a static analysis tool using LLVM/Clang to identify functions in C/C++ code that exhibit data parallelism, i.e. performing the same operation on independent data elements. The tool should emit a diagnostic message for each function demonstrating data parallelism.

Key Tasks
- Introduce a new command-line option (e.g., -analyze-data-parallelism) to enable the data parallelism analysis.
- Define what constitutes data parallelism (e.g., independent iterations of loops).
- Determine patterns and indicators of data parallelism in LLVM IR.
- Identify loops and determine if iterations are independent (i.e., no data dependencies).
- Emit diagnostics for functions with data parallelism, explaining the detected data parallelism.

Example(s)
Data Parallelism 
```
void compute(float* data, int N) {
    for (int i = 0; i < N; ++i) {
        data[i] = sin(data[i]) + cos(data[i]);
    }
}
```

No Data Parallelism
```
void process(float* data, int N) {
    for (int i = 0; i < N; ++i) {
        if (data[i] > 0) {
           data[i] = sqrt(data[i]);
        }
    }
}
```

Expected Output
`Function 'compute' has data parallelism: - Independent loop iterations detected at line 42`

**Comment by zwuis** [comment]
I'm not sure I fully understand the motivation behind this, so I don't have a strong opinion on where to place this tool. If it's intended to be a tool under clang-tools-extra, I'd suggest posting an RFC to https://discourse.llvm.org/ first.

**Comment by HITESH-S-P** [comment]
Thanks for the suggestion! I've now posted an RFC on LLVM Discourse to gather feedback on the proposal and the appropriate integration point.

https://discourse.llvm.org/t/rfc-detecting-data-parallel-loop-patterns-in-c-c-source-using-clang-llvm-analysis/91424

I'll wait for the community's feedback there before proceeding further and will update this issue/PR accordingly.
